### PR TITLE
Compatible data fields

### LINKED ISSUE #1480: compatible_data field(s) in current_law_policy.json
Now that the cps.csv file is much further along, I am opening an issue to renew discussion of adding a compatible_data field in current_law_policy.json. 

In #1074 I suggested:

> Add a "compatible_data" attribute to each parameter after the "notes" field. 
> 
> The acceptable values that I anticipate in the short to medium term are: 
>          - "taxdata_puf", which would refer to the PUF based datafile produced by TaxData.  
>          - "taxdata_cps", which would refer to the CPS based datafile produced by TaxData.
>          - "taxcalc_filings", which would refer to the Luca-based taxpayer inputted data that @zrisher is developing. 
>         - any comma-separated combination of those three, such as "taxdata_puf, taxdata_cps, taxcalc_filings".  
> 
> We need this attribute because not every parameter will be relevant to every data source. For some parameters, all of the relevant variables from the data source might be zero. Implementing reforms to these parameters will not influence the results. Given that our users implement reforms to many provisions at once, many will not notice that there is a problem.  I expect this will lead to quite a few silly mistakes and some resentment. > 

Based on a conversation with @martinholmer, I now believe that we would want to make an explicit rule about what "compatible" means, as there will be gray areas. One possible rule would be, "does this parameter affect results with this given dataset". This is something that we could even test relatively simply, albeit with significant computational expense (as @zrisher notes below).

Here is the rest of @zrisher's response to this proposal from #1074. 

> It will be a bit of extra maintenance work. When changing the way policy is coded, a contributor must remember to update this field if usable data changes, and the relationships may not be easy to trace. These fields could also be affected by changes to any of the listed data sets. It's also computationally expensive to enforce via automated testing. However, the value prop you described is definitely there.

### PR DESCRIPTION
**Pull request by MattHJensen**
Resolves #1480.

### CODE REVIEW

**Comment by MattHJensen** [comment]
Significant unresolved problems with this PR:

- duplicates (more) code in `test_cpscsv.py` and `test_pufcsv.py`
- tests take 10+ minutes to run. 
- tests are currently marked 'pre_release' which means that we could update `TESTING.md` to tell everyday contributors to run `py.test -m "not requires_pufcsv and not pre_release" -n4` or  `py.test -m "not pre_release" -n4`, but that is wordy and more importantly would place a large burden on the Tax-Calculator maintainer who issues the releases. 

I will continue thinking about how to resolve these problems, but I wanted to post a [WIP] PR so others could see the direction this is headed in and chime in with any suggestions.

**Comment by martinholmer** [comment]
Matt said:
> Significant unresolved problems with this PR:
>
> - duplicates (more) code in test_cpscsv.py and test_pufcsv.py

Consider combining the puf and cos tests into a single parametrized function and put that function in its own test file.

**Comment by MattHJensen** [comment]
@martinholmer said:

> Consider combining the new puf and cps tests into a single parametrized function and put that function in its own test file.

Thanks, I will do that.

**Comment by martinholmer** [comment]
@MattHJensen has been failing the check-in tests in pull request #1614 for some weeks, then thought he needed to upgrade to Python 3.6 in order to avoid errors having to do with the different string types in Python 2.7 and Python 3.6+.   Here was a typical error:
```
        for pname in allparams:
            param = allparams[pname]
            max_listed = param['range']['max']
            # Handle links to other params or self
>           if isinstance(max_listed, unicode):
E           NameError: name 'unicode' is not defined
taxcalc/tests/test_cpscsv.py:199: NameError
```
The easiest way to eliminate these kinds of errors is to do what all the other unit tests do.  Here's an example for `test_calculate.py`:
```
    import six
    .
    .
    .
    params = Calculator.read_json_param_objects(reform_json, assump_json)
    assert isinstance(params, dict)
    doc = Calculator.reform_documentation(params)
    assert isinstance(doc, six.string_types)
```
There is no need to upgrade to Python 3.6; just handle 2.7 vs 3.6 differences using the six package, which has been part of the development environment for years.

**Comment by MattHJensen** [comment]
> The easiest way to eliminate these kinds of errors is to do what all the other unit tests do [and use six].

Thanks @martinholmer, a recent commit adopts this approach.

**Comment by MattHJensen** [comment]
This PR is ready for review. Thanks in advance. 

cc @andersonfrailey @martinholmer @hdoupe @GoFroggyRun

**Comment by andersonfrailey** [comment]
Just finished reviewing and this looks good to me.

**Comment by martinholmer** [comment]
@MattHJensen,  I hoping to review/test pull request #1614 later today (Wed, 11/22).

**Comment by martinholmer** [comment]
Pull request #1614 adds a new `taxcalc/tests/test_....py` file that begins this way:
```
"""
Tests of the compatible_data field in current_law_policy.json.
"""
# CODING-STYLE CHECKS:
# pep8 --ignore=E402 test_pufcsv.py
# pylint --disable=locally-disabled test_pufcsv.py

import os
import json
import difflib
import pytest
import numpy as np
import pandas as pd
import copy
import six
# pylint: disable=import-error
from taxcalc import Policy, Records, Calculator


@pytest.mark.requires_pufcsv
@pytest.mark.pre_release
@pytest.mark.parametrize('puftest', [True, False])
def test_compatible_data(puftest, tests_path, cps_subsample, puf_subsample):
    """
    """

    clppath = os.path.join(tests_path, '..', 'current_law_policy.json')
    pfile = open(clppath, 'r')
    allparams = json.load(pfile)
    pfile.close()
    assert isinstance(allparams, dict)
```
If you really want to include this file among those that are subjected to the coding style, then you need to edit the top CODING STYLE comments and fix the code so that it doesn't generate any pylint warnings.  Here's what I get:
```
iMac:tests mrh$ pep8 --ignore=E402 test_compatible_data.py
iMac:tests mrh$ pylint --disable=locally-disabled test_compatible_data.py
************* Module test_compatible_data
C: 73, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C: 76, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:133, 0: Wrong continued indentation (add 1 space).
                                 c_xx.records.combined) *
                                 ^| (bad-continuation)
C:136, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:142, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C: 23, 0: Empty function docstring (empty-docstring)
R: 23, 0: Too many local variables (25/15) (too-many-locals)
R: 23, 0: Too many branches (24/12) (too-many-branches)
R: 23, 0: Too many statements (76/50) (too-many-statements)
W: 10, 0: Unused import difflib (unused-import)
W: 12, 0: Unused numpy imported as np (unused-import)
W: 13, 0: Unused pandas imported as pd (unused-import)
C: 14, 0: standard import "import copy" should be placed before "import pytest" (wrong-import-order)
```

The execution time for this command `py.test -m pre_release -n4` on my iMac is about 4.6 minutes.

@MattHJensen

**Comment by MattHJensen** [comment]
> If you really want to include this file among those that are subjected to the coding style, then you need to edit the top CODING STYLE comments and fix the code so that it doesn't generate any pylint warnings. 

Thanks for pointing this out @martinholmer, I will take a look at it in upcoming days. 

> The execution time for this command py.test -m pre_release -n4 on my iMac is about 4.6 minutes.

This is similar to what I am getting. I am not sure how to speed this up w/o sacrificing the test coverage, but I am very welcome to anyone's suggestions.

**Comment by hdoupe** [comment]
I opened a [PR](https://github.com/MattHJensen/tax-calculator/pull/8) on @MattHJensen 's branch.  In the PR, I tried add more parallelization to these tests.  I was able to get the run time down to 215 seconds which still isn't great.  Maybe fresh eyes will see something that I missed.

**Comment by hdoupe** [comment]
@MattHJensen Whups, that's a pep-8 error. Sorry about that.

**Comment by MattHJensen** [comment]
@hdoupe, that's ok. I need to add a couple of commits anyways.

**Comment by hdoupe** [comment]
@MattHJensen what's the status on this PR?

**Comment by MattHJensen** [comment]
@hdoupe, this is ready for additional review. I am removing the WIP label.

File: taxcalc/tests/test_compatible_data.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    # Get taxcalc objects from tc_objs fixture
+    p_xx, rec_xx, c_xx, puftest = tc_objs
+
+    # These parameters are exempt because they are not active under
+    # current law and activating them would deactive other parameters.
+    exempt = ['_CG_ec', '_CG_reinvest_ec_rt']
+
+    # assert len(allparams_i) == 1
+    # pname = list(allparams_i.keys())[0]
+    # param = allparams_i[pname]
```
Comment (hdoupe) [inline taxcalc/tests/test_compatible_data.py]: @MattHJensen you may want to remove these commented out lines.

**Comment by hdoupe** [comment]
@MattHJensen Looks good to me.

File: taxcalc/tests/test_compatible_data.py (hunk 1 above)
Comment (MattHJensen) [reply taxcalc/tests/test_compatible_data.py]: Done. Thanks.

**Comment by MattHJensen** [comment]
Thanks very much for the review and commits @hdoupe.

**Comment by hdoupe** [comment]
No problem. Nice work!

**Comment by martinholmer** [comment]
@MattHJensen,  About four weeks ago I made [this comment on 1614](https://github.com/open-source-economics/Tax-Calculator/pull/1614#issuecomment-346470925), but things haven't changed on this front.
Here is what I get when I do the checking described in our [coding documentation](https://github.com/open-source-economics/Tax-Calculator/blob/master/CODING.md):
```
iMac2:tests mrh$ pep8 --ignore=E402 test_compatible_data.py 
iMac2:tests mrh$ pylint --disable=locally-disabled test_compatible_data.py 
************* Module test_compatible_data
test_compatible_data.py:103: [C0325(superfluous-parens), ] Unnecessary parens after 'print' keyword
test_compatible_data.py:106: [C0325(superfluous-parens), ] Unnecessary parens after 'print' keyword
test_compatible_data.py:180: [C0330(bad-continuation), ] Wrong continued indentation (add 1 space).
                                 c_xx.records.combined) *
                                 ^|
test_compatible_data.py:183: [C0325(superfluous-parens), ] Unnecessary parens after 'print' keyword
test_compatible_data.py:189: [C0325(superfluous-parens), ] Unnecessary parens after 'print' keyword
test_compatible_data.py:31: [W0109(duplicate-key), reform_xx] Duplicate key '_CTC_new_refund_limited' in dictionary
test_compatible_data.py:72: [W0621(redefined-outer-name), sorted_param_names] Redefining name 'allparams' from outer scope (line 64)
test_compatible_data.py:82: [W0621(redefined-outer-name), allparams_batch] Redefining name 'allparams' from outer scope (line 64)
test_compatible_data.py:82: [W0621(redefined-outer-name), allparams_batch] Redefining name 'sorted_param_names' from outer scope (line 72)
test_compatible_data.py:86: [C0103(invalid-name), allparams_batch] Invalid variable name "ix"
test_compatible_data.py:94: [W0621(redefined-outer-name), tc_objs] Redefining name 'reform_xx' from outer scope (line 25)
test_compatible_data.py:118: [W0621(redefined-outer-name), test_compatible_data] Redefining name 'allparams_batch' from outer scope (line 81)
test_compatible_data.py:118: [W0621(redefined-outer-name), test_compatible_data] Redefining name 'tc_objs' from outer scope (line 94)
test_compatible_data.py:117: [W0621(redefined-outer-name), test_compatible_data] Redefining name 'reform_xx' from outer scope (line 25)
test_compatible_data.py:117: [W0621(redefined-outer-name), test_compatible_data] Redefining name 'allparams' from outer scope (line 64)
test_compatible_data.py:117: [C0112(empty-docstring), test_compatible_data] Empty function docstring
test_compatible_data.py:117: [R0913(too-many-arguments), test_compatible_data] Too many arguments (6/5)
test_compatible_data.py:117: [R0914(too-many-locals), test_compatible_data] Too many local variables (24/15)
test_compatible_data.py:123: [W0612(unused-variable), test_compatible_data] Unused variable 'p_xx'
test_compatible_data.py:117: [R0912(too-many-branches), test_compatible_data] Too many branches (22/12)
test_compatible_data.py:117: [R0915(too-many-statements), test_compatible_data] Too many statements (60/50)
test_compatible_data.py:8: [W0611(unused-import), ] Unused import os
test_compatible_data.py:9: [W0611(unused-import), ] Unused import json
test_compatible_data.py:10: [W0611(unused-import), ] Unused import difflib
test_compatible_data.py:13: [W0611(unused-import), ] Unused pandas imported as pd
test_compatible_data.py:14: [C0411(wrong-import-order), ] standard import "import copy" should be placed before "import pytest"

------------------------------------------------------------------
Your code has been rated at 7.14/10 (previous run: 7.14/10, +0.00)
```
7.14 out of 10 is a pretty low score.  And some of the warnings look like real logic problems.
Also, the CODING comments at the top of the file are incorrect.

I'd be happy to review this pull request again later when its in better shape.

**Comment by MattHJensen** [comment]
I edited the coding comments and dealt with what seemed to be the most important pylint warnings. The following warnings are remaining but don't warrant holding up this PR by my judgement (happy for others to disagree). 
```

C: 97, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:100, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:181, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:187, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
W: 66,23: Redefining name 'allparams' from outer scope (line 58) (redefined-outer-name)
W: 76,29: Redefining name 'allparams' from outer scope (line 58) (redefined-outer-name)
W: 76,40: Redefining name 'sorted_param_names' from outer scope (line 66) (redefined-outer-name)
C: 80, 4: Invalid variable name "ix" (invalid-name)
W: 88,21: Redefining name 'reform_xx' from outer scope (line 20) (redefined-outer-name)
W:112,34: Redefining name 'allparams_batch' from outer scope (line 75) (redefined-outer-name)
W:112,25: Redefining name 'tc_objs' from outer scope (line 88) (redefined-outer-name)
W:111,66: Redefining name 'reform_xx' from outer scope (line 20) (redefined-outer-name)
W:111,55: Redefining name 'allparams' from outer scope (line 58) (redefined-outer-name)
R:111, 0: Too many arguments (6/5) (too-many-arguments)
R:111, 0: Too many local variables (24/15) (too-many-locals)
W:121, 4: Unused variable 'p_xx' (unused-variable)
R:111, 0: Too many branches (22/12) (too-many-branches)
R:111, 0: Too many statements (60/50) (too-many-statements)

-----------------------------------
Your code has been rated at 7.93/10
```

File: taxcalc/tests/test_compatible_data.py [hunk 2]
```diff
@@ -3,15 +3,10 @@
 """
 # CODING-STYLE CHECKS:
 # pep8 --ignore=E402 test_pufcsv.py
```
Comment (martinholmer) [inline taxcalc/tests/test_compatible_data.py]: I don't understand the meaning of this comment.

**Comment by martinholmer** [comment]
The following pylint warnings look scary to me, what's going on in your code?
```
W: 66,23: Redefining name 'allparams' from outer scope (line 58) (redefined-outer-name)
W: 76,29: Redefining name 'allparams' from outer scope (line 58) (redefined-outer-name)
W: 76,40: Redefining name 'sorted_param_names' from outer scope (line 66) (redefined-outer-name)
W: 88,21: Redefining name 'reform_xx' from outer scope (line 20) (redefined-outer-name)
W:112,34: Redefining name 'allparams_batch' from outer scope (line 75) (redefined-outer-name)
W:112,25: Redefining name 'tc_objs' from outer scope (line 88) (redefined-outer-name)
W:111,66: Redefining name 'reform_xx' from outer scope (line 20) (redefined-outer-name)
W:111,55: Redefining name 'allparams' from outer scope (line 58) (redefined-outer-name)
```

**Comment by hdoupe** [comment]
It seems like pylint doesn't like how we used the py.test fixture functions.  For example, the fixture `allparams` is defined in line 58 and is used as an argument in the `sorted_all_params` fixture.  I think that's what the `redefined-outer-name` comment means.

I see that some of the fixture functions were re-named to resolve this issue.  One example is in [test_taxcalcio.py](https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/tests/test_taxcalcio.py#L15).  @MattHJensen You could modify this example to be:
```
@pytest.fixture(scope="module", name="allparams")
def fixture_allparams():
    """
    Get current law parameters
    """
    return Policy.default_data(metadata=True)


@pytest.fixture(scope="module", name="sorted_param_names")
def fixture_sorted_param_names(allparams):
    """
    Fixture for storing a sorted parameter list
    """
    return sorted(list(allparams.keys()))
```

My opinion is that the current implementation is cleaner and more straightforward.  However, I'm happy to hear other people's thoughts.

**Comment by MattHJensen** [comment]
Thanks very much @hdoupe for the clear explanation. I also do not feel strongly about whether to prepend "fixture_" to the beginning of fixture function names. @martinholmer, did @hdoupe's explanation answer your question about the pylint warnings? (Thanks again to both of you for your review and help with this PR).

**Comment by martinholmer** [comment]
@MattHJensen asked in #1614:
> did @hdoupe's explanation answer your question about the pylint warnings?

Yes.

**Pull request merged**